Settings: Flag project as deactivated and hide from tools' view

davidlatwe opened the pull request:
## Feature

Able to switch project state between *active* and *inactive* in Studio Settings.

The project in *inactive* state will not be shown in tools like :
* Launcher
* Library Loader
* Standalone Publisher
* Project Manager
* Local Settings
* Sync Server

## Usage

To mark project as *inactive* :

1. Go to `Projects` tab in Studio Settings and select the project that you want to deactivate.
2. Go to `project_anatomy/attributes/active` and toggle it.
3. Save the settings

![855f991b-acbd-4f33-9f9d-8590b1a453db](https://user-images.githubusercontent.com/3357009/133541462-12426c49-9eb1-4f1e-a4fa-f77193900da1.gif)

### Tips
* The current selected *project* will remain in the list view when it just got deactivated, it will be hidden when you switch to another project settings.
* You can view all projects by checking the "Show Inactive Project" box. This check-box only exists in Studio Settings.


## Notes

* This PR depends on https://github.com/pypeclub/avalon-core/pull/368
* The Sync Server is not yet tested in my end.

tokejepsen commented:
Great stuff! Maybe we could have a confirmation box when using the archive switch rather than a separate line edit @iLLiCiTiT ?

iLLiCiTiT commented on openpype/settings/entities/root_entities.py:
```diff
@@ -755,6 +755,35 @@ def defaults_dir(self):
+
+        # Evaluate project archiving flag
+        #
+        archive_confirm = anatomy.get("attributes", {}).get("archive_confirm")
```
Do not set unset keys by default. That way the key won't be marked as modified (which is because have missing key) and won't show you that project has missing set values...

iLLiCiTiT commented on openpype/tools/settings/settings/widgets.py:
```diff
@@ -612,11 +612,39 @@ def mouseReleaseEvent(self, event):
     project_changed = QtCore.Signal()
 
-    def __init__(self, parent):
+    ProjectSortRole = QtCore.Qt.UserRole + 10
```
Don't use came case for attributes and don't set these constants used in different code in class. It is better to define roles as constants at header part.

davidlatwe replied:
Should be resolved by 0bf84ad

iLLiCiTiT commented on openpype/tools/settings/settings/widgets.py:
```diff
@@ -612,11 +612,39 @@ def mouseReleaseEvent(self, event):
+
+        index = self.sourceModel().index(source_row, 0, source_parent)
+        is_active = bool(index.data(self.filterRole()))
+        is_selected = bool(index.data(ProjectListWidget.ProjectSelectedRole))
```
I quite don't get why selection is important? All projects should disappear when checkbox change value...

davidlatwe replied:
This was for keeping current selected project visible in list when it just got deactivated.

Since the only way for user to know which project is being edited is from the selection of that list view, so I thought it would be better not to hide the project that just got deactivated from list as soon as the "Save" is clicked and processed, until the user changed the selection to another project settings.

iLLiCiTiT commented:
I would maybe change `archived` to `active`. The validation does not make much sense as you are not triggering any callbacks. It's just a flag to show/hide projects for artists plus you may accidentally "unarchive" project when you save project settings on "archived" project because confirmation won't be passed (through API).

Ad active: Also you may want to have inactive projects by default which does not make sense with having `archived` by default or just want pause project but not archive yet.

mkolar commented:
I'd also stay away from `archived` considering studios tend to use that as and information when project data is actually archived and not easily accessible. `Active` is quite nice, with it being on by default. But it also allows as @iLLiCiTiT mentioned for other workflows. For example preparing a project structure and some date before marking it as active and showing to the artists.

davidlatwe commented:
All good points, I have changed the attrib `archive` to `active`.

> It's just a flag to show/hide projects for artists plus you may accidentally "unarchive" project when you save project settings on "archived" project because confirmation won't be passed (through API).

That's true. I removed the confirmation for now as it's also outdated after the `archive` attrib is changed.

But after the confirmation is removed, code like this (e.g. in avalon-core code base `AvalonMongoDB.projects`) may not work as expect if the default setting `active` is changed to `false` :
```python
for doc in project_docs:
    if not doc["data"].get("active", True):
        # the default value here is not following studio default setting
        continue
    ...
```

Not sure how to solve this properly, as it may also make sense if one like the project's default active state to `False` ?

For new project, we could make the project manager to always set the `data.active` in database with studio default when creating new one. But for the existing old projects, how do we handle that ?

iLLiCiTiT commented:
> But after the confirmation is removed, code like this (e.g. in avalon-core code base AvalonMongoDB.projects) may not work as expect if the default setting active is changed to false

I don't think it's used in a way that it would change behavior or affect anything.

> But for the existing old projects, how do we handle that ?

Since there is (in avalon-core) default value of "active" set to `True` (`project_doc["data"].get("active", True)`) it should be fine as all projects without the key are proposed as active.

> How about scan and update project documents that doesn't have data.active and set it to true on pype tray launch ?

Not in tray like a script before/after launch that's not good idea as it would slow down all tray applications if studio has 30 projects it would query 30 times on each tray launch without doing anything. It could be done by script (out of tray), through settings or could modify `AvalonMongoDB.projects` to set the key to `True` if is not set or value is `None`. The last one is most easiest but is forcing to have that key on project...

iLLiCiTiT commented on openpype/tools/settings/settings/widgets.py:
```diff
@@ -730,17 +797,37 @@ def refresh(self):
+
+            for doc in self.dbcon.projects(
+                    projection={"name": 1, "data.active": 1},
+                    only_active=self._only_active
```
This should be always `False`. You want have them in project model all the time but filter them with proxy model.

iLLiCiTiT replied:
Oh it;s because of SyncServer?

davidlatwe replied:
Ouch, this currently still being shared with the Sync-Server. Have thought about letting Sync-Server to have it's own project list like you suggested in chat, but that requires a bit more code change to do so. So as current workaround, I make that as optional and fix to `False` once Sync-Server has it's own widget. Maybe in another PR ?

davidlatwe replied:
> Oh it;s because of SyncServer?

XD, didn't see the update. Yes :D

davidlatwe replied:
Ah, forgot to mention. Local Settings' project list widget also inherited this.

iLLiCiTiT commented on openpype/tools/settings/settings/widgets.py:
```diff
@@ -640,11 +679,27 @@ def __init__(self, parent):
+            layout.addWidget(archived_chk, 0)
+            layout.addSpacing(5)
+
+            archived_chk.stateChanged.connect(self.on_archive_vis_changed)
```
Change variable names with `archived` to `active` and method callbacks too...

davidlatwe commented:
> Since there is (in avalon-core) default value of "active" set to True (project_doc["data"].get("active", True)) it should be fine as all projects without the key are proposed as active.

Yeah, but if the default value in Studio Settings gets change to `false`, the code won't be align with the settings. Or this behavior should be considered as edge case ?

iLLiCiTiT commented:
> Or this behavior should be considered as edge case ?

Since project document is existing studio defaults of anatomy data are "ignored" so it is ok...

davidlatwe commented:
Thanks folks, I have also updated the top comment. 🚀

mkolar commented on openpype/settings/entities/schemas/projects_schema/schemas/schema_anatomy_attributes.json:
```diff
@@ -69,6 +69,11 @@
+        {
+            "type": "boolean",
+            "key": "active",
+            "label": "Is Project Active"
```
```suggestion
            "label": "Active Project"
```

The only suggestion I have is to rename this label to a noun rather than a question. Other than that. All good

mkolar reviewed (changes requested):
Actually I spoke too soon. Sync queue is showing no projects at all with this PR. @davidlatwe get in touch with @kalisp he should be able to point to to a simple site sync setup with local storage so you can sort it out.

iLLiCiTiT commented on openpype/tools/settings/settings/widgets.py:
```diff
@@ -612,11 +618,35 @@ def mouseReleaseEvent(self, event):
+
+    def __init__(self, *args, **kwargs):
+        super(ProjectListSortFilterProxy, self).__init__(*args, **kwargs)
+        self._enable_filter = True
```
```suggestion
        self._enable_filter = False
```

iLLiCiTiT commented on openpype/tools/settings/settings/widgets.py:
```diff
@@ -640,11 +679,27 @@ def __init__(self, parent):
+        if only_active:
+            inactive_chk = None
+        else:
+            inactive_chk = QtWidgets.QCheckBox(" Show Inactive Projects ")
```
Use parenting on widgets for proper style propagation.
```suggestion
            inactive_chk = QtWidgets.QCheckBox(" Show Inactive Projects ", self)
```

davidlatwe replied:
Hmmm, I tried this but doesn't have any difference.

iLLiCiTiT replied:
I know but it may differ if we had different styles for checkbox than default.

iLLiCiTiT commented on openpype/tools/settings/settings/widgets.py:
```diff
@@ -640,11 +679,27 @@ def __init__(self, parent):
         layout.addWidget(label_widget, 0)
         layout.addWidget(project_list, 1)
 
+        if only_active:
+            inactive_chk = None
```
Option n2 for SyncServer fix
```suggestion
        if only_active:
            inactive_chk = None
            project_proxy.set_filter_enabled(False)
```

davidlatwe commented:
Thanks @iLLiCiTiT for taking a stab on fixing SyncServer ❤️ 

However I found an issue :
1. I deactivate a project
2. open up SyncServer and find that inactive project still showing in the list
3. select that inactive project in SyncServer
4. entire Pype tray crashed

As for Launcher, LibraryLoader, StandalonePublisher, LocalSettings and ProjectManager, they all seems find in a few quick tests.

mkolar commented:
>However I found an issue :

Oh good testing. I literally just tested too and didn't notice that

mkolar commented:
But I can confirm that it indeed crashes in the scenario that @davidlatwe described

iLLiCiTiT commented:
So in the end @kalisp will have to look into that :)

davidlatwe commented:
Here's the full log and GIF that shows the bug
https://github.com/davidlatwe/OpenPype/pull/1#issuecomment-921072845

mkolar commented:
Task linked: [PYPE-1427 Pype Projects - Deleting & Archiving](https://app.clickup.com/t/6658547/PYPE-1427)

kalisp commented:
Fixed Site Sync issue in https://github.com/davidlatwe/OpenPype/pull/2

mkolar commented:
@kalisp @davidlatwe apologies. I've prematurely merged this thinking the fix was already part of it here. 

Petr could you please make the same PR to this main repo and we'll make a hotfix asap with it. Thanks.
